### ISSUE TITLE
php-phpcs throws abnormal exit/fails to lint

### ISSUE THREAD

**Issue by tigerhawkvok** [flycheck/flycheck#1241]
Just installed everything off of the latest versions.

PHPCS is throwing a weird exit code, with flycheck reporting the following:

```
Suspicious state from syntax checker php-phpcs: Flycheck checker php-phpcs returned non-zero exit code 1, but its output contained no errors:
```

It then renders and XML doc of style errors, eg, 

```xml
<?xml version="1.0" encoding="UTF-8"?>
<checkstyle version="2.8.1">
<file name="FILE.php">
for a file comment" source="PEAR.Commenting.FileComment.WrongStyle"/>
 <error line="7" column="1" severity="error" message="Perl-style comments are not allowed. Use &quot;// Comment.&quot; or &quot;/* comment */&quot; instead." source="PEAR.Commenting.InlineComment.WrongStyle"/>
```

I can get the bad exit code message to go away [by changing the exit code in `CLI.php`](https://github.com/squizlabs/PHP_CodeSniffer/blob/2.8.1/CodeSniffer/CLI.php#L117), but then Flycheck doesn't notify of the style errors.

It's possible Flycheck needs to set this flag: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Configuration-Options#ignoring-errors-when-generating-the-exit-code

Yet even when I manually edit the file to return a zero exit code, Flycheck still doesn't notify me of coding standard issues

The result of `flycheck-compile php-phpcs` after manually changing the exit code:

```
-*- mode: compilation; default-directory: "path/to/file/" -*-
Compilation started at Fri Apr 21 12:23:23

"phpcs" "--report=checkstyle" "-q" "--stdin-path=FILE.php" < "FILE.php"
<?xml version="1.0" encoding="UTF-8"?>
<checkstyle version="2.8.1">
[...]
</file>
</checkstyle>

Compilation finished at Fri Apr 21 12:23:24
```

**Comment by fmdkdd** [comment]
Thank you for the report.

> PHPCS is throwing a weird exit code, with flycheck reporting the following:

Most tools that report errors follow the convention to return non-zero if any errors are present, and zero if there are no errors to report.  Of course, if something is wrong with the tool *itself* (e.g., it crashed), it will usually also return non-zero.  Thus, when a tool returns non-zero, Flycheck will expect to parse some errors.  If there are no errors to parse, then the invocation is "suspicious" because either something is wrong with the tool (it crashed), or something is wrong with Flycheck's parsing of the tool output.

So, if you invoke `phpcs` manually on your file and it reports style errors with a non-zero exit, that's normal.  If, for the same file, Flycheck does *not* report the errors in your buffer, then Flycheck behavior is at fault.

Can you provide a small PHP file on which PHPCS reports errors that are not picked up by Flycheck?

**Comment by cpitclaudel** [comment]
Thanks @fmdkdd for answering so quickly.  I'll add that the XML in the OP seems invalid, which would explain why Flycheck gets confused: it sees a non-zero exit code (which is fine, since there is an error), but then it looks for the error and cannot find it (because the XML is invalid), and so it complains: the tool claimed there were errors (thorough its exit code), but then no error could be found in the XML output.

Was that the actual XML returned by phpcs?

**Comment by tigerhawkvok** [comment]
Got it. It was win/libxml2 thing.

For other users, [this libxml2 archive dropped in my Emacs directory worked for me](http://xmlsoft.org/sources/win32/64bit/libxml2-2.9.3-win32-x86_64.7z), but [these dependencies](https://github.com/cpitclaudel/company-coq/issues/20#issuecomment-249938296) should work too.

I *also* had to set `C:\Windows\Temp` to give user `Everyone` **Full Control**

[Google Drive mirror](https://drive.google.com/file/d/0B6LJvOj7KC-cTXQ2eDJLbDAtT2c/view?usp=sharing)

Past-you-guys [pointed me in the right direction](https://github.com/flycheck/flycheck/issues/491#issuecomment-56356534).

May not be remiss to put the windows dependency link in the "Windows Support" section.

**Comment by fmdkdd** [comment]
@tigerhawkvok Wow, okay.  So this is a windows thing?  Flycheck fails to parse the XML without libxml2 installed, and there are no errors saying so?

Glad to know it works for you.

**Comment by tigerhawkvok** [comment]
@fmdkdd it was several things. FIrst the tempfile failed to write for permissions reasons, so I was getting `fclose() expects parameter 1 to be handle, not boolean`; then fixing the permissions, it was just very briefly flashing the libxml2 thing before popping up a blob of text. I only saw the libxml2 thing as an emacs error when trying on a different, slower machine (very briefly).

**Comment by fmdkdd** [comment]
@tigerhawkvok So there is a warning, althoug transient.  That's comforting ;)

Happy hacking.
